pkg/alertmanager/multitenant.go

```diff
@@ -435,11 +436,17 @@ func (am *MultitenantAlertmanager) setConfig(cfg alerts.AlertConfigDesc) error {
 		am.alertmanagersMtx.Unlock()
 	} else if am.cfgs[cfg.User].RawConfig != cfg.RawConfig || hasTemplateChanges {
 		level.Info(am.logger).Log("msg", "updating new per-tenant alertmanager", "user", cfg.User)
+		activeBefore := existing.IsActive()
 		// If the config changed, apply the new one.
 		err := existing.ApplyConfig(cfg.User, userAmConfig, rawCfg)
 		if err != nil {
 			return fmt.Errorf("unable to apply Alertmanager config for user %v: %v", cfg.User, err)
 		}
+
+		// If we went from paused to active, we need to re-add the user-regitry.
+		if !activeBefore && existing.IsActive() {
+			am.alertmanagerMetrics.addUserRegistry(cfg.User, existing.registry)
```

Comment:
If I'm not missing anything, this should not be correct. The problem is that `existing.registry` is the previous Prometheus registry and not a new one. This means that the registered metrics are there since before `deleteUserRegistry()` has been called and this would lead to counters/histograms/summaries to be double tracked as soon as you re-add it via `addUserRegistry()`. Why? Because when we call `deleteUserRegistry()` the last value of all counters/histograms/summaries is "recorded" in the `UserRegistries`. When you add back the same registry via `addUserRegistry()`, the metrics in the registry will be summed up to the last value of the metrics when `deleteUserRegistry()` was called, but if the registry is the same you will end up with double countings.

Assuming we fix this issue, the only benefit I see of this PR is let **gauge metrics** values (only gauges) disappear from aggregated metrics when `deleteUserRegistry()`. If this is the desiderata of this PR, then to fix it we should use a new registry each time we call `addUserRegistry()`.

_Am I missing anything?_

Reply:
Thanks, Marco for pointing this out - I think you're right on the soft deletion part. We need new registry.

My intention was to remove _all_ metrics, but under the current circumstances (where we can't distinguish between a user moving away from a shard or being removed completely from the system) if this is the best we can do (without incurring in more complexity) - I'll take it.